### PR TITLE
(EAI-98): Longer conversation length (not dependent on context window)

### PR DESCRIPTION
**Pull request by mongodben**
Jira: https://jira.mongodb.org/browse/EAI-98

## Changes

- Add `FilterPreviousMessages` function
- Add implementations of `FilterPreviousMessage` function:
  - `filterOnlySystemPrompt()`: default. only include the initial system prompt
  - `makeFilterNPreviousMessages()`: Constructor of filter to include initial system prompt + N previous messages in conversation
- Add FilterPreviousMessages config to the conversations router
- Add implementation of `FilterPreviousMessages` to the MongoDB Chatbot server. Includes increasing conversation max number of user messages to 50.
- Document addition of `FilterPreviousMessages`

## Notes

- In the future we can consider increasing the max number of user messages. The current 50 was an arbitrary "biggish" number that won't break the server.

### CODE REVIEW

File: packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.ts [hunk 1]
```diff
@@ -71,6 +74,8 @@ export function makeAddMessageToConversationRoute({
   generateUserPrompt,
   maxInputLengthCharacters = DEFAULT_MAX_INPUT_LENGTH,
   maxUserMessagesInConversation = DEFAULT_MAX_USER_MESSAGES_IN_CONVERSATION,
+  // TODO: use
```
Comment (cbush) [inline packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.ts:77]: ?

Comment (mongodben) [reply packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.ts:77]:
```suggestion
```

File: packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.test.ts [hunk 2]
```diff
@@ -128,6 +129,39 @@ describe("POST /conversations/:conversationId/messages", () => {
       origin,
     });
   });
+  test("uses previous message filter", async () => {
+    const mockMsg = {
+      role: "user",
+      // This corresponds to ~130 tokens
+      content:
+        "In a quaint village nestled among verdant hills, a whimsical inventor, Eleanor, toiled in her sun-dappled workshop. Every day, she crafted extraordinary gadgets, each more intricate than the last. One morning, as the village awoke to a chorus of songbirds, Eleanor unveiled her latest creation: a clockwork bird with iridescent feathers. This mechanical marvel could mimic any tune it heard. The villagers gathered, marveling as the bird sang melodies from distant lands. Meanwhile, beyond the rolling meadows, a curious fox watched, its amber eyes reflecting a world where nature and Eleanor's inventions coexisted in harmonious wonder.",
```
Comment (nlarew) [inline packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.test.ts:137]: Eerily close to this AI generated story! https://storybird.ai/library/children-stories/the-mystery-of-good-view-village

**Review by nlarew — APPROVED** [review]

File: packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.test.ts (hunk 2 above)
Comment (mongodben) [reply packages/mongodb-chatbot-server/src/routes/conversations/addMessageToConversation.test.ts:137]: we were probably using similar ChatGPT prompts lol.

i hadn't seen storybird.ai before. that's neat. very very similar to [FableFusion](https://github.com/bpmutter/fablefusion), my first gen AI app. i wasn't able to figure out the image generation part (this was back in ~dec '22). curious how they handled it.

**Pull request merged**